amqp/src/main/scala/akka/stream/alpakka/amqp/impl/AmqpAsyncFlow.scala

```diff
@@ -0,0 +1,332 @@
+    val callback = getAsyncCallback[(DeliveryTag, Boolean)] {
+      case (tag: DeliveryTag, multiple: Boolean) => confirmCallback(tag, multiple)
+    }
+    new ConfirmCallback { // cant use function literal because it doesn't work with 2.11
+      override def handle(tag: DeliveryTag, multiple: Boolean): Unit = callback.invoke((tag, multiple))
+    }
+  }
+
+  private def onConfirmation(tag: DeliveryTag, multiple: Boolean): Unit = {
+    log.debug("Received confirmation for deliveryTag {} (multiple={}).", tag, multiple)
+
+    val dequeued: Iterable[AwaitingMessage[T]] = dequeueAwaitingMessages(tag, multiple)
+
+    dequeued.foreach(m => cancelTimer(m.tag))
+
+    pushOrEnqueueResults(
+      dequeued.map(m => WriteResult.confirmed(m.passThrough))
+    )
+  }
+
+  private def onRejection(tag: DeliveryTag, multiple: Boolean): Unit = {
+    log.debug("Received rejection for deliveryTag {} (multiple={}).", tag, multiple)
+
+    val dequeued: Iterable[AwaitingMessage[T]] = dequeueAwaitingMessages(tag, multiple)
+
+    dequeued.foreach(m => cancelTimer(m.tag))
+
+    pushOrEnqueueResults(
+      dequeued.map(m => WriteResult.rejected(m.passThrough))
+    )
+  }
+
+  private def pushOrEnqueueResults(results: Iterable[WriteResult[T]]): Unit = {
+    results.foreach(result => {
+      if (exitQueue.nonEmpty || !isAvailable(out)) {
+        log.debug("Message {} queued for downstream push.", result)
+        exitQueue.enqueue(result)
+      } else {
+        log.debug("Pushing {} downstream.", result)
+        push(out, result)
+      }
+    })
+    if (isFinished) closeStage()
+  }
+
+  override def postStop(): Unit = {
+    promise.tryFailure(new RuntimeException("Stage stopped unexpectedly."))
+    super.postStop()
+  }
+
+  override def onFailure(ex: Throwable): Unit = {
+    promise.tryFailure(ex)
+    onFailure(ex)
+  }
+
+  def dequeueAwaitingMessages(tag: DeliveryTag, multiple: Boolean): Iterable[AwaitingMessage[T]]
+
+  def enqueueMessage(tag: DeliveryTag, passThrough: T): Unit
+
+  def messagesAwaitingDelivery: Int
+
+  def noAwaitingMessages: Boolean
+
+  setHandler(
+    in,
+    new InHandler {
+
+      override def onPush(): Unit = {
+        val message: WriteMessage[T] = grab(in)
+        val tag = publish(message)
+
+        scheduleOnce(tag, confirmationTimeout)
+        enqueueMessage(tag, message.passThrough)
+        if (messagesAwaitingDelivery + exitQueue.size < bufferSize && !hasBeenPulled(in)) tryPull(in)
+      }
+
+      override def onUpstreamFailure(ex: Throwable): Unit = {
+        upstreamException = Some(ex)
+        if (isFinished)
+          closeStage()
+        else
+          log.debug("Received upstream failure signal - stage will be failed when all buffered messages are processed")
+
+      }
+
+      override def onUpstreamFinish(): Unit =
+        if (noAwaitingMessages && exitQueue.isEmpty) {
+          promise.success(Done)
```

Comment:
`tryFailure`, but `success`, why not `trySuccess`? From reading code it is not entirely clear how often each method is called... could it be possible to be called more than once?

Reply:
`onUpstreamFinish` is called only once, thats why it should be safe to call just `success`. `tryFailure` is used in `postStop` and `onFailure`. `postStop` is executed even if you close stage manually, so it's more like a fallback if promise is not resolved in any correct way. `onFailure` is called in a connection shutdown handler (see `AmqpConnectorLogic`) so I'm not sure if it would be safe to call `failure`.